// @glimpse/glimpse-server/release/src/resources/HudEmbeddedResource.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var express = require("express");
var path = require("path");
var glimpse_common_1 = require("@glimpse/glimpse-common");
var Resource = (function () {
    function Resource(server) {
        var _this = this;
        this.name = 'hud';
        this.uriTemplate = 'main.js?hash={hash}';
        this.templateName = 'hudScriptTemplate';
        this.type = 'client';
        this.invoke = function (req, res, next) {
            if (!_this.sentHudImpressionTelemetry) {
                _this.sentHudImpressionTelemetry = true;
                var props = { clientIP: glimpse_common_1.RequestHelper.getIp(req) };
                _this.telemetryService.sendEvent(glimpse_common_1.TelemetryEvents.NODE_SERVER_HUD_RESOURCE_ACCESSED, props, {});
            }
            return _this.innerInvoke(req, res, next);
        };
        var useDevBuild = server.providers.configSettings.get('use.embedded.dev.builds', false);
        var basePath = path.join(__dirname, '../../resources/embedded/hud');
        var location = useDevBuild ? path.join(basePath, 'dev') : path.join(basePath, 'prod');
        this.telemetryService = server.providers.telemetryService;
        this.innerInvoke = express.static(location);
    }
    return Resource;
}());
exports.Resource = Resource;
