// @glimpse/glimpse-server/release/src/resources/VersionCheckResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var request = require("request");
var glimpse_common_1 = require("@glimpse/glimpse-common");
var Resource = (function () {
    function Resource(server) {
        this.name = 'version-check';
        this.uriTemplate = '{?currentVersion}';
        this.type = 'client';
        this.errorReportingService = server.providers.errorReportingService;
    }
    Resource.prototype.invoke = function (req, res) {
        var that = this;
        var url = 'https://registry.npmjs.org/@glimpse%2Fglimpse';
        request.get({ url: url, json: true }, function (err, resp, body) {
            if (!err) {
                // forward body through as our response
                res.json(body);
            }
            else {
                // make sure we report the error
                that.errorReportingService.reportError(glimpse_common_1.createVersionCheckError(err));
                // still need to request
                res.status(500).send('Verion check failed: ' + err);
            }
        });
    };
    return Resource;
}());
exports.Resource = Resource;
