// @glimpse/glimpse-agent-node/release/messaging/MessagePublisherHttp.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var GlimpseServerRequest_1 = require("./GlimpseServerRequest");
var ContextManagerProvider_1 = require("./ContextManagerProvider");
var MessagePublisherHttp = (function () {
    function MessagePublisherHttp(resourceProvider) {
        this.resourceProvider = resourceProvider;
    }
    MessagePublisherHttp.prototype.publishMessages = function (messages) {
        var context = ContextManagerProvider_1.ContextManagerProvider.getContextManager().currentContext();
        var resources = this.resourceProvider.getResourceDefinitions(context);
        var options = {
            uri: resources[MessagePublisherHttp.messageIngressTemplate],
            method: 'POST',
            json: true,
            body: messages
        };
        GlimpseServerRequest_1.glimpseServerRequest(options.uri, options, function onResponse(err) {
            if (err) {
                throw err;
            }
            for (var i = 0; i < messages.length; i++) {
                messages[i].sent = true;
            }
        });
    };
    ;
    MessagePublisherHttp.messageIngressTemplate = 'messageIngressTemplate';
    return MessagePublisherHttp;
}());
exports.MessagePublisherHttp = MessagePublisherHttp;
