// @glimpse/glimpse-agent-node/release/tracing/module_instrumentors/DebugModuleInstrumentor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DebugProxy_1 = require("./DebugProxy");
/**
 * The module instrumentor for the [`debug`](https://www.npmjs.com/package/debug) package.
 */
var DebugModuleInstrumentor = (function () {
    function DebugModuleInstrumentor() {
    }
    Object.defineProperty(DebugModuleInstrumentor.prototype, "supportedModules", {
        get: function () {
            return {
                'debug': '>=2.0 <2.7'
            };
        },
        enumerable: true,
        configurable: true
    });
    DebugModuleInstrumentor.prototype.setAgent = function (agent) {
        this.agent = agent;
    };
    DebugModuleInstrumentor.prototype.enableInstrumentation = function (moduleInfo) {
        var debugProxy = new DebugProxy_1.default(this.agent.broker, this.agent.providers.consoleManager, this.agent.providers.contextManager, this.agent.providers.messageConverter, this.agent.providers.stackHelper);
        var newModule = debugProxy.proxyModule(moduleInfo.originalModule, moduleInfo.version);
        // Right now we don't do any in-depth checking to see if the module couldn't
        // be enabled for some reason, so we hard return true here
        return {
            isEnabled: true,
            newModule: newModule
        };
    };
    return DebugModuleInstrumentor;
}());
exports.DebugModuleInstrumentor = DebugModuleInstrumentor;
exports.default = DebugModuleInstrumentor;
