// @glimpse/glimpse-common/src/errors/TelemetryErrorReportingService.ts
import * as chalk from 'chalk';
import * as util from 'util';

import { IErrorReportingService } from './IErrorReportingService';
import { IGlimpseError, GlimpseErrorClass, GlimpseErrorSeverity } from './IGlimpseError';
import { ITelemetryService, IProperties, IMeasurements } from '../telemetry/TelemetryService';
import { TelemetryEvents } from '../telemetry/TelemetryEvents';
import { getSlugForErrorCode } from './ErrorCodes';
import { PackageHelper } from '../common/PackageHelper';

export class TelemetryErrorReportingService implements IErrorReportingService {

    private telemetryService: ITelemetryService;
    private sentDependencies = false;

    constructor(telemetryService: ITelemetryService) {
        this.telemetryService = telemetryService;
    }

    public reportError(error: IGlimpseError, ...params): void {

        if (this.telemetryService.isEnabled) {

            if (!this.sentDependencies) {
                this.sendDependencies();
            }

            const measurements: IMeasurements = {};
            const properties: IProperties = {
                // fill in a 'stack' property on the properties object
                stack: new Error().stack,
                severity: GlimpseErrorSeverity[error.severity],
                errorClass: GlimpseErrorClass[error.errorClass],
                errorCode: '' + error.errorCode,
                errorSlug: getSlugForErrorCode(error.errorCode),
                // NOTE: Messages may be chalk-formatted so we strip that before being reported.
                message: chalk.stripColor(util.format(error.message, ...params))
            };

            this.telemetryService.sendEvent(TelemetryEvents.ERROR, properties, measurements);
        }
    }

    private sendDependencies() {
        if (this.telemetryService.isEnabled && !this.sentDependencies) {
            const appPackageJsonPath = PackageHelper.instance.findAppPackageJsonPath();
            const dependencies = PackageHelper.instance.getDependencies(appPackageJsonPath);
            const devDependencies = PackageHelper.instance.getDevDependencies(appPackageJsonPath);

            const measurements: IMeasurements = {};
            const properties: IProperties = {
                dependencies: JSON.stringify(dependencies),
                devDependencies: JSON.stringify(devDependencies)
            };
            this.telemetryService.sendEvent(TelemetryEvents.DEPENDENCIES, properties, measurements);
            this.sentDependencies = true;
        }
    }
}
