// @glimpse/glimpse-agent-node/messaging/MessagePublisherHttp.ts
'use strict';

import { glimpseServerRequest } from './GlimpseServerRequest';
import { IMessagePublisher } from './IMessagePublisher';
import { IResourceProvider } from '../configuration/IResourceProvider';
import { ContextManagerProvider } from './ContextManagerProvider';

export class MessagePublisherHttp implements IMessagePublisher {
    private static messageIngressTemplate = 'messageIngressTemplate';

    public constructor(private resourceProvider: IResourceProvider) {
    }

    public publishMessages(messages) {
        const context = ContextManagerProvider.getContextManager().currentContext();
        const resources = this.resourceProvider.getResourceDefinitions(context);
        const options = {
            uri: resources[MessagePublisherHttp.messageIngressTemplate],
            method: 'POST',
            json: true,
            body: messages
        };

        glimpseServerRequest(options.uri, options, function onResponse(err) {
            if (err) {
                throw err;
            }

            for (let i = 0; i < messages.length; i++) {
                messages[i].sent = true;
            }
        });
    };
}
