// @panter/vue-i18next/lib/component.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = {
  name: 'i18next',
  functional: true,
  props: {
    tag: {
      type: String,
      default: 'span'
    },
    path: {
      type: String,
      required: true
    }
  },
  render: function render(h, _ref) {
    var props = _ref.props,
        data = _ref.data,
        children = _ref.children,
        parent = _ref.parent;

    var i18next = parent.$i18n;
    if (!i18next) {
      return h(props.tag, data, children);
    }

    var path = props.path;

    var REGEXP = i18next.i18next.services.interpolator.regexp;
    var format = i18next.t(path, { interpolation: { prefix: '#$?', suffix: '?$#' } });
    var tchildren = [];

    format.split(REGEXP).reduce(function (memo, match, index) {
      var child = void 0;
      if (index % 2 === 0) {
        if (match.length === 0) return memo;

        child = match;
      } else {
        child = children[parseInt(match, 10)];
      }

      memo.push(child);
      return memo;
    }, tchildren);

    return h(props.tag, data, tchildren);
  }
};
module.exports = exports['default'];
