// @panter/vue-i18next/lib/install.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Vue = undefined;
exports.install = install;

var _component = require('./component');

var _component2 = _interopRequireDefault(_component);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Vue = exports.Vue = void 0; /* eslint-disable import/no-mutable-exports */
function install(_Vue) {
  if (install.installed) {
    return;
  }
  install.installed = true;

  exports.Vue = Vue = _Vue;

  Vue.mixin({
    computed: {
      $t: function $t() {
        var _this = this;

        if (this._i18nOptions) {
          var _i18nOptions = this._i18nOptions,
              lng = _i18nOptions.lng,
              namespaces = _i18nOptions.namespaces;

          var fixedT = this.$i18n.i18next.getFixedT(lng, namespaces);
          return function (key, options) {
            return fixedT(key, options, _this.$i18n.i18nLoadedAt);
          };
        }
        return function (key, options) {
          return _this.$i18n.i18next.t(key, options, _this.$i18n.i18nLoadedAt);
        };
      }
    },

    beforeCreate: function beforeCreate() {
      var options = this.$options;
      if (options.i18n) {
        this.$i18n = options.i18n;
      } else if (options.parent && options.parent.$i18n) {
        this.$i18n = options.parent.$i18n;
      }

      if (this.$i18n && this.$options.i18nOptions) {
        var _$options$i18nOptions = this.$options.i18nOptions.lng,
            lng = _$options$i18nOptions === undefined ? null : _$options$i18nOptions;
        var namespaces = this.$options.i18nOptions.namespaces;

        namespaces = namespaces || this.$i18n.i18next.options.defaultNS;
        if (typeof namespaces === 'string') namespaces = [namespaces];
        this.$i18n.i18next.loadNamespaces(namespaces);

        this._i18nOptions = { lng: lng, namespaces: namespaces };
      } else if (this.$i18n && options.parent && options.parent._i18nOptions) {
        this._i18nOptions = options.parent._i18nOptions;
      }
    }
  });

  Vue.component(_component2.default.name, _component2.default);
}
